src/water_quality/mapping/algorithms.py

```diff
@@ -12,73 +13,315 @@
+        },
+        "tm": {
+            "red": ("tm03", 660.0),
+            "nir": ("tm04", 830.0),
+            "swir": ("tm05", 1650.0),
+        },
+        "tm_agm": {
+            "red": ("tm03_agm", 660.0),
+            "nir": ("tm04_agm", 830.0),
+            "swir": ("tm05_agm", 1650.0),
+        },
+    }
+    if instrument not in CENTRAL_WAVELENGTHS.keys():
+        log.error(
+            f"Invalid instrument '{instrument}'. FAI will be calculated as zero."
+        )
+        return xr.zeros_like(ds[list(ds.data_vars)[0]]).rename("FAI")
+    else:
+        # Extract band names (vars) and central wavelengths (l) for the
+        # selected instrument
+        inst_bands = CENTRAL_WAVELENGTHS[instrument]
+        red_band, l_red = inst_bands["red"]
+        nir_band, l_nir = inst_bands["nir"]
+        swir_band, l_swir = inst_bands["swir"]
+
+        # Compute FAI
+        # FAI = Observed_NIR - Interpolated_NIR
+        # Observerd NIR is the Rayleigh-corrected reflectance in the NIR
+        # Interpolated_NIR = Red + (SWIR - Red) * (NIR_wl - Red_wl) / (SWIR_wl - Red_wl)
+        # Interpolated_NIR is a linear baseline formed by the red and SWIR bands.
+        # It follows the convention Slope-intercept form:
+        # y = mx+b, where m is the slope of the line and b is the y-intercept.
+
+        # Calculate the slope coefficient (m) for the baseline
+        # m = (NIR_wl - Red_wl) / (SWIR_wl - Red_wl)
+        m = (l_nir - l_red) / (l_swir - l_red)
+
+        # Calculate the baseline reflectance at NIR wavelength
+        interpolated_nir = ds[red_band] + (ds[swir_band] - ds[red_band]) * m
+        # Scale by 10000 assuming input reflectance is in 0-10000 range
+        fai = ds[nir_band] - interpolated_nir
+        fai.name = "FAI"
+        return fai / 10000
+
+
+def geomedian_FAI(ds: xr.Dataset) -> xr.Dataset:
+    """
+    Calculate the FAI (Floating Algae Index) across multiple instruments
+    and produce a combined weighted mean FAI for water pixels.
+
+    Parameters
+    ----------
+    ds : xr.Dataset
+        Input dataset containing the spectral bands.
+
+    Returns
+    -------
+    xr.DataArray
+        The updated input xarray Dataset with new bands:
+        - '{instrument}_fai' for each processed instrument (masked).
+        - 'agm_fai' (the final weighted average FAI, masked).
+    """
+
+    reference_mean = {
+        "fai": {"msi_agm": 0.0970, "oli_agm": 0.1015, "tm_agm": 0.0962},
+    }
+    # threshold = {
+    #     'fai' : {'msi_agm': 0.05   , 'oli_agm' : 0.05   , 'tm_agm': 0.05},
+    #    }
+    fai_threshold = 0.05
+    # Keep this order for consistent processing.
+    geomedian_instruments = ["msi_agm", "oli_agm", "tm_agm"]
+    # The instrument to use as the scaling reference
+    reference_inst = "msi_agm"
+
+    # Initialization for Weighted Average
+    mean_fai_weighted_sum = None
+    agm_count_total = None
+
+    for inst_agm in geomedian_instruments:
```

Comment:
not sure if there's actual performance impact, but in general would to good to eliminate loop....